### DISCUSSION TITLE
graphql Optional filters

### DISCUSSION THREAD

**Discussion by cpt-n3mo** [directus/directus#10477] [Archive]
Good day,

Using GraphQL, how would one use a optional filter, as a example

```graphql
query myProducts($catagory: String!, $make: String, $subCatagory: [String]){
    myProducts(filter:{
      catagory:{ name:{ _eq: $catagory}}
      make:{ name: {_eq: $make }}
      subCatagory:{ mySubCatagories_id: {name: {_in: $subCatagory }}}
    }){
      name
      image{
        id
      }
      smallText
    }
  }
```
I would like the query to run even if the variables are not specified,. it now returns a empty array, as the filter is still used even when there empty.
to what should i set my variables so that the filter is ignored..?
using null does not seem to work..

kind regards

**Comment by cpt-n3mo** [comment]
not sure if this is best practice, but this seems to work,
I made the filter object into a variable, this allowed me to build the filter dynamically.
this solved the problem for me..

**Reply by hillport** [reply]
Hey,
I can't seem to get the syntax right for this. Would be great if you could share how you declared it.
Thanks ;)

**Reply by cpt-n3mo** [reply]
Good Day!

I hope this little bit of code will help you get going.. 
  
A computed value for the filter,  make what ever logic you need to compose your reactive filter
basically the filter is the same as you would make it in Graphql query, just now inside a JavaScript object, leave out the elements you dont need, i push the values depending on condition to a empty _and array inside a filter object.

```
productFilter(){
  return {filter: { _and: [ 
  {cat: { name: { _eq: this.activeTabName } } },
  {make:{ name: {_eq: this.make }}},
  {subCatagory:{ mySubCatagories_id: {name: {_in: this.subCatagory }}}}
] }}
}

```
then Apollo will look something like this.
```
myProducts: {
      prefetch: false,
      query: myProducts,
      variables() {
        return this.productFilter;
      },
    },
```
then the query should look something like this..
notice that the type of the filter variable is that of the `<Collection Name>_filter`  type

```
query myProducts($filter: myProducts_filter){
    myProducts(filter: $filter){
      id
    }
  } 
```

im not sure if you are using something to compose your queries.. Im using Altair now, and it makes query building quite easy i pulls the docs nicely so you can get intellisense on the whole filter process / query, this makes creating a query match easier.. :)
 
cheers!

**Reply by hillport** [reply]
That's exactly what I was struggling with, the filter variable type. Couldn't get it to work you saved my sanity. Everything works now !
Thanks a bunch for the tips too, I'll check that up ;)
